client/state/media/reducer.js

```diff
@@ -0,0 +1,70 @@
+/**
+ * External dependencies
+ */
+import { combineReducers } from 'redux';
+import { reject, keyBy, map } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import { createReducer, keyedReducer } from 'state/utils';
+import {
+	MEDIA_ITEMS_RECEIVE,
+	MEDIA_FILE_UPLOAD,
+	MEDIA_FILE_UPLOAD_FAILURE,
+	MEDIA_FILE_UPLOAD_SUCCESS,
+	MEDIA_FILE_UPLOADS_ENQUEUE
+} from 'state/action-types';
+import { createTransientMedia } from './utils';
+
+/**
+ * Returns the updated upload queue state after an action has been dispatched.
+ * The state is an array of pairs: siteId, file.
+ *
+ * @param  {Object} state  Current state
+ * @param  {Object} action Action payload
+ * @return {Object}        Updated state
+ */
+export const uploadQueue = createReducer( [], {
+	[ MEDIA_FILE_UPLOADS_ENQUEUE ]: ( state, { siteId, files } ) => [
+		...state,
+		...files.map( ( file ) => [ siteId, file ] )
+	],
+	[ MEDIA_FILE_UPLOAD_FAILURE ]: ( state, { file } ) => reject( state, ( [ , queued ] ) => queued === file ),
+	[ MEDIA_FILE_UPLOAD_SUCCESS ]: ( state, { file } ) => reject( state, ( [ , queued ] ) => queued === file )
```

Comment:
>in some of my code I have found it helpful to include a comment in destructured lists indicating what's missing

Yeah, makes sense. If ESLint doesn't complain about it, would it make sense to just include the destructured variable uncommented?